checks.d/splunk_topology.py

```diff
@@ -0,0 +1,190 @@
+"""
+    StackState.
+    Splunk topology extraction
+"""
+
+# 3rd party
+import requests
+from urllib import quote
+import time
+
+# project
+from checks import AgentCheck, CheckException
+
+
+class SavedSearch:
+    def __init__(self, saved_search):
+        self.name = saved_search['name']
+        self.element_type = saved_search['element_type']
+        self.parameters = saved_search['parameters']
+
+
+class InstanceConfig:
+    def __init__(self, instance, init_config):
+        default_timeout = init_config.get('default_timeout', 5)
+        max_retry_count = init_config.get('max_retry_count', 3)
+        seconds_between_retries = init_config.get('seconds_between_retries', 1)
+        polling_interval = init_config.get('polling_interval', 15)
+
+        self.base_url = instance['url']
+        self.username = instance['username']
+        self.password = instance['password']
+        self.timeout = float(instance.get('timeout', default_timeout))
+        self.max_retry_count = int(instance.get('max_retry_count', max_retry_count))
+        self.seconds_between_retries = int(instance.get('seconds_between_retries', seconds_between_retries))
+        self.polling_interval = int(instance.get('polling_interval', polling_interval))
+
+    def get_auth_tuple(self):
+        return self.username, self.password
+
+
+class Instance:
+    INSTANCE_TYPE = "splunk"
+
+    def __init__(self, instance, init_config):
+        self.instance_config = InstanceConfig(instance, init_config)
+        self.saved_searches = [SavedSearch(saved_search) for saved_search in instance['saved_searches']]
+        self.instance_key = {
+            "type": self.INSTANCE_TYPE,
+            "url": self.instance_config.base_url
+        }
+        self.tags = instance.get('tags', [])
+        self.last_successful_poll = 0
+
+    def should_poll(self, time_seconds):
+        return self.last_successful_poll == 0 or time_seconds >= self.last_successful_poll + self.instance_config.polling_interval
+
+
+class SplunkTopology(AgentCheck):
+    # SERVICE_CHECK_NAME = "splunk.topology_information"
+    # service_check_needed = True
+
+    def __init__(self, name, init_config, agentConfig, instances=None):
+        super(SplunkTopology, self).__init__(name, init_config, agentConfig, instances)
+        # Data to keep over check runs, keyed by instance url
+        self.instance_data = dict()
+
+    def check(self, instance):
+        if 'url' not in instance:
+            raise CheckException('Splunk topology instance missing "url" value.')
+
+        if instance["url"] not in self.instance_data:
+            self.instance_data[instance["url"]] = Instance(instance, self.init_config)
```

Comment:
Im not sure if its a good idea to require instances to have unique urls. E.g. at KPN they probably have one url for splunk. This means they have to store all there searches inside one instance, and have the same polling interval for them etc. Allthough we know allready that some information is updated once every half hour and other once every day. If this is the idea of instances perhaps we need more granularity inside an instance, to define different polling intervals.